// @livechat/redux-utils/lib/index.js
'use strict';

exports.__esModule = true;
exports.createPropsActions = exports.createPropsSelectors = exports.createPropsReducer = exports.getOrIdentity = exports.createReducer = exports.createActionCreator = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.combineReducers = combineReducers;

var _getOr = require('lodash/fp/getOr');

var _getOr2 = _interopRequireDefault(_getOr);

var _omit = require('lodash/fp/omit');

var _omit2 = _interopRequireDefault(_omit);

var _map = require('lodash/fp/map');

var _map2 = _interopRequireDefault(_map);

var _values = require('lodash/fp/values');

var _values2 = _interopRequireDefault(_values);

var _reduce = require('lodash/fp/reduce');

var _reduce2 = _interopRequireDefault(_reduce);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createActionCreator = exports.createActionCreator = function createActionCreator(type, actionCreator) {
	var wrappedActionCreator = function wrappedActionCreator() {
		return {
			type: type,
			payload: actionCreator.apply(undefined, arguments)
		};
	};
	wrappedActionCreator.getActionName = function () {
		return type;
	};
	// (wrappedActionCreator: any).toString = (): string => (type);
	return wrappedActionCreator;
};

var createReducer = exports.createReducer = function createReducer(intialState, reducersMap) {
	return function reducer(state, action, globalState) {
		var currentState = state || intialState;
		if (reducersMap.hasOwnProperty(action.type)) {
			return reducersMap[action.type](currentState, action.payload, globalState);
		}
		return currentState;
	};
};

function combineReducers(reducers) {
	var reducerKeys = Object.keys(reducers);
	var finalReducers = {};
	for (var index = 0; index < reducerKeys.length; index++) {
		var key = reducerKeys[index];
		if (typeof reducers[key] === 'function') {
			finalReducers[key] = reducers[key];
		}
	}
	var finalReducerKeys = Object.keys(finalReducers);

	return function combination() {
		var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
		var action = arguments[1];
		var globalState = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : state;

		var hasChanged = false;
		var nextState = {};
		for (var _index = 0; _index < finalReducerKeys.length; _index++) {
			var _key = finalReducerKeys[_index];
			var reducer = finalReducers[_key];
			var previousStateForKey = state[_key];
			var nextStateForKey = reducer(previousStateForKey, action, globalState);
			nextState[_key] = nextStateForKey;
			hasChanged = hasChanged || nextStateForKey !== previousStateForKey;
		}
		return hasChanged ? nextState : state;
	};
}

var getOrIdentity = exports.getOrIdentity = function getOrIdentity(path, state) {
	return (0, _getOr2.default)(state, path, state);
};

var capitalizeFirstLetter = function capitalizeFirstLetter(text) {
	return text.charAt(0).toUpperCase() + text.slice(1);
};

var createPropsReducer = exports.createPropsReducer = function createPropsReducer(propName) {
	var _ref;

	return _ref = {}, _ref['SET_PROPERTY_' + propName.toUpperCase()] = function undefined(state, payload) {
		var _extends2;

		return _extends({}, state, (_extends2 = {}, _extends2[propName] = payload.data, _extends2));
	}, _ref['REMOVE_PROPERTY_' + propName.toUpperCase()] = function undefined(state) {
		return _extends({}, (0, _omit2.default)(propName, state));
	}, _ref;
};

var createPropsSelectors = exports.createPropsSelectors = function createPropsSelectors(propName) {
	var _ref2;

	return _ref2 = {}, _ref2['getProperty' + capitalizeFirstLetter(propName)] = function undefined(state) {
		return state[propName];
	}, _ref2;
};

var createPropsActions = exports.createPropsActions = function createPropsActions(propName) {
	var _ref3;

	return _ref3 = {}, _ref3['setProperty' + capitalizeFirstLetter(propName)] = createActionCreator('SET_PROPERTY_' + propName.toUpperCase(), function (data) {
		return {
			data: data
		};
	}), _ref3['removeProperty' + capitalizeFirstLetter(propName)] = createActionCreator('REMOVE_PROPERTY_' + propName.toUpperCase(), function () {
		return {
			propName: propName
		};
	}), _ref3;
};
