// @livechat/livechat-visitor-sdk/lib/actions/chatActions.js
'use strict';

exports.__esModule = true;
exports.setVisitorData = exports.rateChat = exports.typingIndicator = exports.visitorQueued = exports.newFile = exports.visitorDataSetError = exports.visitorDataSet = exports.newAgent = exports.chatEnded = exports.newMessage = exports.chatStarted = undefined;

var _reduxUtils = require('@livechat/redux-utils');

var CHAT_STARTED = 'CHAT_STARTED';

var NEW_MESSAGE = 'NEW_MESSAGE';
var CHAT_ENDED = 'CHAT_ENDED';
var NEW_AGENT = 'NEW_AGENT';
var VISITOR_DATA_SET = 'VISITOR_DATA_SET';
var VISITOR_DATA_SET_ERROR = 'VISITOR_DATA_SET_ERROR';
var NEW_FILE = 'NEW_FILE';
var VISITOR_QUEUED = 'VISITOR_QUEUED';
var TYPING_INDICATOR = 'TYPING_INDICATOR';
var RATE_CHAT = 'RATE_CHAT';
var SET_VISITOR_DATA = 'SET_VISITOR_DATA';

var chatStarted = exports.chatStarted = (0, _reduxUtils.createActionCreator)(CHAT_STARTED, function (_ref) {
	var securedToken = _ref.securedToken;
	return { securedToken: securedToken };
});
var newMessage = exports.newMessage = (0, _reduxUtils.createActionCreator)(NEW_MESSAGE, function (_ref2) {
	var id = _ref2.id,
	    timestamp = _ref2.timestamp,
	    authorType = _ref2.authorType,
	    authorId = _ref2.authorId,
	    text = _ref2.text,
	    chatId = _ref2.chatId,
	    isMobile = _ref2.isMobile,
	    customId = _ref2.customId;
	return {
		id: id,
		timestamp: timestamp,
		authorType: authorType,
		authorId: authorId,
		text: text,
		chatId: chatId,
		isMobile: isMobile,
		customId: customId
	};
});
var chatEnded = exports.chatEnded = (0, _reduxUtils.createActionCreator)(CHAT_ENDED, function () {
	return null;
});
var newAgent = exports.newAgent = (0, _reduxUtils.createActionCreator)(NEW_AGENT, function (_ref3) {
	var avatarUrl = _ref3.avatarUrl,
	    jobTitle = _ref3.jobTitle,
	    name = _ref3.name,
	    id = _ref3.id;
	return {
		avatarUrl: avatarUrl,
		jobTitle: jobTitle,
		name: name,
		id: id
	};
});
var visitorDataSet = exports.visitorDataSet = (0, _reduxUtils.createActionCreator)(VISITOR_DATA_SET, function (_ref4) {
	var visitorData = _ref4.visitorData,
	    requestId = _ref4.requestId;
	return {
		visitorData: visitorData,
		requestId: requestId
	};
});
var visitorDataSetError = exports.visitorDataSetError = (0, _reduxUtils.createActionCreator)(VISITOR_DATA_SET_ERROR, function (_ref5) {
	var error = _ref5.error,
	    requestId = _ref5.requestId;
	return {
		error: error,
		requestId: requestId
	};
});
var newFile = exports.newFile = (0, _reduxUtils.createActionCreator)(NEW_FILE, function (_ref6) {
	var id = _ref6.id,
	    timestamp = _ref6.timestamp,
	    authorType = _ref6.authorType,
	    authorId = _ref6.authorId,
	    url = _ref6.url,
	    contentType = _ref6.contentType,
	    size = _ref6.size;
	return {
		id: id,
		timestamp: timestamp,
		authorType: authorType,
		authorId: authorId,
		url: url,
		size: size,
		contentType: contentType
	};
});
var visitorQueued = exports.visitorQueued = (0, _reduxUtils.createActionCreator)(VISITOR_QUEUED, function (_ref7) {
	var numberInQueue = _ref7.numberInQueue,
	    waitingTime = _ref7.waitingTime;
	return {
		numberInQueue: numberInQueue,
		waitingTime: waitingTime
	};
});
var typingIndicator = exports.typingIndicator = (0, _reduxUtils.createActionCreator)(TYPING_INDICATOR, function (_ref8) {
	var authorId = _ref8.authorId,
	    isTyping = _ref8.isTyping;
	return {
		authorId: authorId,
		isTyping: isTyping
	};
});
var rateChat = exports.rateChat = (0, _reduxUtils.createActionCreator)(RATE_CHAT, function (_ref9) {
	var rate = _ref9.rate,
	    comment = _ref9.comment;
	return { rate: rate, comment: comment };
});
var setVisitorData = exports.setVisitorData = (0, _reduxUtils.createActionCreator)(SET_VISITOR_DATA, function (_ref10) {
	var visitorData = _ref10.visitorData,
	    requestId = _ref10.requestId;
	return {
		visitorData: visitorData,
		requestId: requestId
	};
});
