// @livechat/livechat-visitor-sdk/lib/actions/propsActions.js
'use strict';

exports.__esModule = true;
exports.propsConstants = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _reduxUtils = require('@livechat/redux-utils');

var propsConstants = exports.propsConstants = {
	LICENSE: 'license',
	GROUP: 'group',
	SESSION_ID: 'sessionId',
	SECURED_TOKEN: 'securedToken',
	CHAT_ID: 'chatId',
	CHAT_STATE: 'chatState',
	CHAT_STATUS: 'chatStatus',
	VISITOR_ID: 'visitorId',
	CUSTOM_VISITOR_ID: 'customVisitorId',
	LAST_MESSAGE_ID: 'lastMessageId',
	CONNECTION_STATUS: 'connectionStatus',
	SNEAK_PEEK: 'sneakPeek',
	TICKET_FORM: 'ticketForm',
	STATIC_CONFIG_VERSION: 'staticConfigVersion',
	SECRET_ID: 'secretId'
};

var propsActions = _extends({}, (0, _reduxUtils.createPropsActions)(propsConstants.LICENSE), (0, _reduxUtils.createPropsActions)(propsConstants.GROUP), (0, _reduxUtils.createPropsActions)(propsConstants.SESSION_ID), (0, _reduxUtils.createPropsActions)(propsConstants.SECURED_TOKEN), (0, _reduxUtils.createPropsActions)(propsConstants.CHAT_ID), (0, _reduxUtils.createPropsActions)(propsConstants.CHAT_STATE), (0, _reduxUtils.createPropsActions)(propsConstants.VISITOR_ID), (0, _reduxUtils.createPropsActions)(propsConstants.CUSTOM_VISITOR_ID), (0, _reduxUtils.createPropsActions)(propsConstants.LAST_MESSAGE_ID), (0, _reduxUtils.createPropsActions)(propsConstants.CHAT_STATUS), (0, _reduxUtils.createPropsActions)(propsConstants.CONNECTION_STATUS), (0, _reduxUtils.createPropsActions)(propsConstants.SNEAK_PEEK), (0, _reduxUtils.createPropsActions)(propsConstants.TICKET_FORM), (0, _reduxUtils.createPropsActions)(propsConstants.STATIC_CONFIG_VERSION), (0, _reduxUtils.createPropsActions)(propsConstants.SECRET_ID));

exports.default = propsActions;
