// @livechat/livechat-visitor-sdk/lib/actions/publicActions.js
'use strict';

exports.__esModule = true;
exports.publicSendTicketForm = exports.publicGetTicketForm = exports.publicDisconnect = exports.publicCloseChat = exports.publicForwardChatTranscript = exports.publicSetSneakPeek = exports.publicRateChat = exports.publicCallbackError = exports.publicGetVisitorData = exports.publicSetVisitorData = exports.publicInit = exports.publicSendMessage = undefined;

var _reduxUtils = require('@livechat/redux-utils');

var PUBLIC_SEND_MESSAGE = 'PUBLIC_SEND_MESSAGE';

var PUBLIC_INIT = 'PUBLIC_INIT';
var PUBLIC_GET_VISITOR_DATA = 'GET_VISITOR_DATA';
var PUBLIC_SET_VISITOR_DATA = 'PUBLIC_SET_VISITOR_DATA';
var PUBLIC_CALLBACK_ERROR = 'PUBLIC_CALLBACK_ERROR';
var PUBLIC_RATE_CHAT = 'PUBLIC_RATE_CHAT';
var PUBLIC_SET_SNEAK_PEEK = 'PUBLIC_SET_SNEAK_PEEK';
var PUBLIC_FORWARD_CHAT_TRANSCRIPT = 'PUBLIC_FORWARD_CHAT_TRANSCRIPT';
var PUBLIC_CLOSE_CHAT = 'PUBLIC_CLOSE_CHAT';
var PUBLIC_DISCONNECT = 'PUBLIC_DISCONNECT';
var PUBLIC_GET_TICKET_FORM = 'PUBLIC_GET_TICKET_FORM';
var PUBLIC_SEND_TICKET_FORM = 'PUBLIC_SEND_TICKET_FORM';
var PUBLIC_DESTROY = 'PUBLIC_DESTROY';

var publicSendMessage = exports.publicSendMessage = (0, _reduxUtils.createActionCreator)(PUBLIC_SEND_MESSAGE, function (_ref) {
	var customId = _ref.customId,
	    text = _ref.text,
	    promise = _ref.promise;
	return {
		customId: customId,
		text: text,
		promise: promise
	};
});

var publicInit = exports.publicInit = (0, _reduxUtils.createActionCreator)(PUBLIC_INIT, function (_ref2) {
	var license = _ref2.license,
	    group = _ref2.group;
	return {
		license: license,
		group: group
	};
});

var publicSetVisitorData = exports.publicSetVisitorData = (0, _reduxUtils.createActionCreator)(PUBLIC_SET_VISITOR_DATA, function (_ref3) {
	var name = _ref3.name,
	    email = _ref3.email,
	    pageUrl = _ref3.pageUrl,
	    pageTitle = _ref3.pageTitle,
	    customProperties = _ref3.customProperties,
	    promise = _ref3.promise;
	return {
		name: name,
		email: email,
		customProperties: customProperties,
		pageUrl: pageUrl,
		pageTitle: pageTitle,
		promise: promise
	};
});

var publicGetVisitorData = exports.publicGetVisitorData = (0, _reduxUtils.createActionCreator)(PUBLIC_GET_VISITOR_DATA, function (_ref4) {
	var promise = _ref4.promise;
	return { promise: promise };
});

var publicCallbackError = exports.publicCallbackError = (0, _reduxUtils.createActionCreator)(PUBLIC_CALLBACK_ERROR, function (_ref5) {
	var errorMessage = _ref5.errorMessage;
	return {
		errorMessage: errorMessage
	};
});

var publicRateChat = exports.publicRateChat = (0, _reduxUtils.createActionCreator)(PUBLIC_RATE_CHAT, function (_ref6) {
	var rate = _ref6.rate,
	    comment = _ref6.comment,
	    promise = _ref6.promise;
	return {
		rate: rate,
		comment: comment,
		promise: promise
	};
});

var publicSetSneakPeek = exports.publicSetSneakPeek = (0, _reduxUtils.createActionCreator)(PUBLIC_SET_SNEAK_PEEK, function (_ref7) {
	var text = _ref7.text,
	    promise = _ref7.promise;
	return {
		text: text,
		promise: promise
	};
});

var publicForwardChatTranscript = exports.publicForwardChatTranscript = (0, _reduxUtils.createActionCreator)(PUBLIC_FORWARD_CHAT_TRANSCRIPT, function (_ref8) {
	var email = _ref8.email,
	    promise = _ref8.promise;
	return {
		email: email,
		promise: promise
	};
});

var publicCloseChat = exports.publicCloseChat = (0, _reduxUtils.createActionCreator)(PUBLIC_CLOSE_CHAT, function (_ref9) {
	var promise = _ref9.promise;
	return { promise: promise };
});

var publicDisconnect = exports.publicDisconnect = (0, _reduxUtils.createActionCreator)(PUBLIC_DISCONNECT, function (_ref10) {
	var promise = _ref10.promise;
	return { promise: promise };
});

var publicGetTicketForm = exports.publicGetTicketForm = (0, _reduxUtils.createActionCreator)(PUBLIC_GET_TICKET_FORM, function (_ref11) {
	var promise = _ref11.promise;
	return { promise: promise };
});
var publicSendTicketForm = exports.publicSendTicketForm = (0, _reduxUtils.createActionCreator)(PUBLIC_SEND_TICKET_FORM, function (_ref12) {
	var name = _ref12.name,
	    email = _ref12.email,
	    subject = _ref12.subject,
	    message = _ref12.message,
	    promise = _ref12.promise;
	return { name: name, email: email, subject: subject, message: message, promise: promise };
});
