// @livechat/livechat-visitor-sdk/es/actions/chatActions.js
import { createActionCreator } from '@livechat/redux-utils';

var CHAT_STARTED = 'CHAT_STARTED';
var NEW_MESSAGE = 'NEW_MESSAGE';
var CHAT_ENDED = 'CHAT_ENDED';
var NEW_AGENT = 'NEW_AGENT';
var VISITOR_DATA_SET = 'VISITOR_DATA_SET';
var VISITOR_DATA_SET_ERROR = 'VISITOR_DATA_SET_ERROR';
var NEW_FILE = 'NEW_FILE';
var VISITOR_QUEUED = 'VISITOR_QUEUED';
var TYPING_INDICATOR = 'TYPING_INDICATOR';
var RATE_CHAT = 'RATE_CHAT';
var SET_VISITOR_DATA = 'SET_VISITOR_DATA';

export var chatStarted = createActionCreator(CHAT_STARTED, function (_ref) {
	var securedToken = _ref.securedToken;
	return { securedToken: securedToken };
});
export var newMessage = createActionCreator(NEW_MESSAGE, function (_ref2) {
	var id = _ref2.id,
	    timestamp = _ref2.timestamp,
	    authorType = _ref2.authorType,
	    authorId = _ref2.authorId,
	    text = _ref2.text,
	    chatId = _ref2.chatId,
	    isMobile = _ref2.isMobile,
	    customId = _ref2.customId;
	return {
		id: id,
		timestamp: timestamp,
		authorType: authorType,
		authorId: authorId,
		text: text,
		chatId: chatId,
		isMobile: isMobile,
		customId: customId
	};
});
export var chatEnded = createActionCreator(CHAT_ENDED, function () {
	return null;
});
export var newAgent = createActionCreator(NEW_AGENT, function (_ref3) {
	var avatarUrl = _ref3.avatarUrl,
	    jobTitle = _ref3.jobTitle,
	    name = _ref3.name,
	    id = _ref3.id;
	return {
		avatarUrl: avatarUrl,
		jobTitle: jobTitle,
		name: name,
		id: id
	};
});
export var visitorDataSet = createActionCreator(VISITOR_DATA_SET, function (_ref4) {
	var visitorData = _ref4.visitorData,
	    requestId = _ref4.requestId;
	return {
		visitorData: visitorData,
		requestId: requestId
	};
});
export var visitorDataSetError = createActionCreator(VISITOR_DATA_SET_ERROR, function (_ref5) {
	var error = _ref5.error,
	    requestId = _ref5.requestId;
	return {
		error: error,
		requestId: requestId
	};
});
export var newFile = createActionCreator(NEW_FILE, function (_ref6) {
	var id = _ref6.id,
	    timestamp = _ref6.timestamp,
	    authorType = _ref6.authorType,
	    authorId = _ref6.authorId,
	    url = _ref6.url,
	    contentType = _ref6.contentType,
	    size = _ref6.size;
	return {
		id: id,
		timestamp: timestamp,
		authorType: authorType,
		authorId: authorId,
		url: url,
		size: size,
		contentType: contentType
	};
});
export var visitorQueued = createActionCreator(VISITOR_QUEUED, function (_ref7) {
	var numberInQueue = _ref7.numberInQueue,
	    waitingTime = _ref7.waitingTime;
	return {
		numberInQueue: numberInQueue,
		waitingTime: waitingTime
	};
});
export var typingIndicator = createActionCreator(TYPING_INDICATOR, function (_ref8) {
	var authorId = _ref8.authorId,
	    isTyping = _ref8.isTyping;
	return {
		authorId: authorId,
		isTyping: isTyping
	};
});
export var rateChat = createActionCreator(RATE_CHAT, function (_ref9) {
	var rate = _ref9.rate,
	    comment = _ref9.comment;
	return { rate: rate, comment: comment };
});
export var setVisitorData = createActionCreator(SET_VISITOR_DATA, function (_ref10) {
	var visitorData = _ref10.visitorData,
	    requestId = _ref10.requestId;
	return {
		visitorData: visitorData,
		requestId: requestId
	};
});
