// @livechat/livechat-visitor-sdk/es/actions/propsActions.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import { createPropsActions } from '@livechat/redux-utils';

export var propsConstants = {
	LICENSE: 'license',
	GROUP: 'group',
	SESSION_ID: 'sessionId',
	SECURED_TOKEN: 'securedToken',
	CHAT_ID: 'chatId',
	CHAT_STATE: 'chatState',
	CHAT_STATUS: 'chatStatus',
	VISITOR_ID: 'visitorId',
	CUSTOM_VISITOR_ID: 'customVisitorId',
	LAST_MESSAGE_ID: 'lastMessageId',
	CONNECTION_STATUS: 'connectionStatus',
	SNEAK_PEEK: 'sneakPeek',
	TICKET_FORM: 'ticketForm',
	STATIC_CONFIG_VERSION: 'staticConfigVersion',
	SECRET_ID: 'secretId'
};

var propsActions = _extends({}, createPropsActions(propsConstants.LICENSE), createPropsActions(propsConstants.GROUP), createPropsActions(propsConstants.SESSION_ID), createPropsActions(propsConstants.SECURED_TOKEN), createPropsActions(propsConstants.CHAT_ID), createPropsActions(propsConstants.CHAT_STATE), createPropsActions(propsConstants.VISITOR_ID), createPropsActions(propsConstants.CUSTOM_VISITOR_ID), createPropsActions(propsConstants.LAST_MESSAGE_ID), createPropsActions(propsConstants.CHAT_STATUS), createPropsActions(propsConstants.CONNECTION_STATUS), createPropsActions(propsConstants.SNEAK_PEEK), createPropsActions(propsConstants.TICKET_FORM), createPropsActions(propsConstants.STATIC_CONFIG_VERSION), createPropsActions(propsConstants.SECRET_ID));

export default propsActions;
