// @livechat/livechat-visitor-sdk/lib/sagas/requests/closeChat.js
'use strict';

exports.__esModule = true;

var _sagaUtils = require('@livechat/saga-utils');

var _serverRequest = require('../../utils/serverRequest');

var _effects = require('redux-saga/effects');

var _reducers = require('../../reducers');

var _marked = /*#__PURE__*/regeneratorRuntime.mark(closeChat);

function closeChat() {
	var license, visitorId, securedToken, environment, closeChatData;
	return regeneratorRuntime.wrap(function closeChat$(_context) {
		while (1) {
			switch (_context.prev = _context.next) {
				case 0:
					_context.next = 2;
					return (0, _effects.select)(_reducers.propsSelectors.getPropertyLicense);

				case 2:
					license = _context.sent;
					_context.next = 5;
					return (0, _effects.call)(_sagaUtils.waitForState, _reducers.propsSelectors.getPropertyVisitorId);

				case 5:
					visitorId = _context.sent;
					_context.next = 8;
					return (0, _effects.select)(_reducers.propsSelectors.getPropertySecuredToken);

				case 8:
					securedToken = _context.sent;
					_context.next = 11;
					return (0, _effects.select)(_reducers.getEnvironment);

				case 11:
					environment = _context.sent;
					closeChatData = {
						environment: environment,
						license: license,
						securedToken: securedToken,
						body: {
							source: 'visitor',
							random_id: String(Math.random()),
							visitor: {
								id: visitorId
							}
						}
					};
					_context.next = 15;
					return (0, _effects.call)(_serverRequest.sendPostToWebserv, 'close_chat', closeChatData);

				case 15:
					return _context.abrupt('return', _context.sent);

				case 16:
				case 'end':
					return _context.stop();
			}
		}
	}, _marked, this);
}

exports.default = closeChat;
