// @livechat/livechat-visitor-sdk/lib/sagas/requests/sendMessage.js
'use strict';

exports.__esModule = true;

var _sagaUtils = require('@livechat/saga-utils');

var _effects = require('redux-saga/effects');

var _serverRequest = require('../../utils/serverRequest');

var _reducers = require('../../reducers');

var _marked = /*#__PURE__*/regeneratorRuntime.mark(sendMessage);

function sendMessage(_ref) {
	var customId = _ref.customId,
	    text = _ref.text;
	var license, visitorId, securedToken, secretId, environment, sendMessageData, sendMessageResponse;
	return regeneratorRuntime.wrap(function sendMessage$(_context) {
		while (1) {
			switch (_context.prev = _context.next) {
				case 0:
					_context.next = 2;
					return (0, _effects.select)(_reducers.propsSelectors.getPropertyLicense);

				case 2:
					license = _context.sent;
					_context.next = 5;
					return (0, _effects.call)(_sagaUtils.waitForState, _reducers.propsSelectors.getPropertyVisitorId);

				case 5:
					visitorId = _context.sent;
					_context.next = 8;
					return (0, _effects.select)(_reducers.propsSelectors.getPropertySecuredToken);

				case 8:
					securedToken = _context.sent;
					_context.next = 11;
					return (0, _effects.select)(_reducers.propsSelectors.getPropertySecretId);

				case 11:
					secretId = _context.sent;
					_context.next = 14;
					return (0, _effects.select)(_reducers.getEnvironment);

				case 14:
					environment = _context.sent;
					sendMessageData = {
						environment: environment,
						license: license,
						securedToken: securedToken,
						secretId: secretId,
						body: {
							author: 'visitor',
							message: text,
							message_id: customId,
							random_id: customId,
							visitor: {
								id: visitorId
							}
						}
					};
					_context.next = 18;
					return (0, _effects.call)(_serverRequest.sendPostToWebserv, 'send_message', sendMessageData);

				case 18:
					sendMessageResponse = _context.sent;

				case 19:
				case 'end':
					return _context.stop();
			}
		}
	}, _marked, this);
}

exports.default = sendMessage;
