// @livechat/livechat-visitor-sdk/es/sagas/requests/closeChat.js
var _marked = /*#__PURE__*/regeneratorRuntime.mark(closeChat);

import { waitForState } from '@livechat/saga-utils';
import { sendPostToWebserv } from '../../utils/serverRequest';
import { call, select } from 'redux-saga/effects';
import { propsSelectors, getEnvironment } from '../../reducers';

function closeChat() {
	var license, visitorId, securedToken, environment, closeChatData;
	return regeneratorRuntime.wrap(function closeChat$(_context) {
		while (1) {
			switch (_context.prev = _context.next) {
				case 0:
					_context.next = 2;
					return select(propsSelectors.getPropertyLicense);

				case 2:
					license = _context.sent;
					_context.next = 5;
					return call(waitForState, propsSelectors.getPropertyVisitorId);

				case 5:
					visitorId = _context.sent;
					_context.next = 8;
					return select(propsSelectors.getPropertySecuredToken);

				case 8:
					securedToken = _context.sent;
					_context.next = 11;
					return select(getEnvironment);

				case 11:
					environment = _context.sent;
					closeChatData = {
						environment: environment,
						license: license,
						securedToken: securedToken,
						body: {
							source: 'visitor',
							random_id: String(Math.random()),
							visitor: {
								id: visitorId
							}
						}
					};
					_context.next = 15;
					return call(sendPostToWebserv, 'close_chat', closeChatData);

				case 15:
					return _context.abrupt('return', _context.sent);

				case 16:
				case 'end':
					return _context.stop();
			}
		}
	}, _marked, this);
}

export default closeChat;
